[analyzer] Add time-trace scopes for high-level analyzer steps

necto opened the pull request:
Specifically, add a scope for 
- each work-list step,
- each entry point,
- each checker run within a step, and
- bug-suppression phase at the end of the analysis of an entry-point.

These scopes add no perceptible run-time overhead when time-tracing is disabled. You can enable it and generate a time trace using the `-ftime-trace=file.json` option.

See also [the RFC](https://discourse.llvm.org/t/analyzer-rfc-ftime-trace-time-scopes-for-steps-and-entry-points/84343).

--
CPP-6065

necto committed:
Add time-trace scopes for each work-list step and each entry point

Also detail:
- each checker run within each step
- bug-suppression at the end of an entry-point analysis

necto committed:
Make code formatter happy

firewave commented:
Does this also work when CSA is called through clang-tidy? The analyzer profiling data in missing in the `--enable-check-profile` output - see #73673?

Xazax-hun reviewed (commented):
I haven't finished reviewing yet but some nits inline.

necto commented:
> Does this also work when CSA is called through clang-tidy? The analyzer profiling data in missing in the `--enable-check-profile` output - see #73673?

As far as I know, it does not work out of the box with clang-tidy. Time tracing works only if you setup a profiler singleton instance with `llvm::timeTraceProfilerInitialize`, and it is invoked from the clang driver (clang/tools/driver/cc1_main.cpp) and some other non-clang tools, but not from clang-tidy. It should be relatively easy to add, but I don't want to extend the scope of this PR.

necto committed:
Less prescriptive language

necto committed:
mention the app name for the screenshot

necto committed:
[NFC] rename kindToStr to getProgramPointKindName

necto committed:
[NFC] Return StringRef instead of a raw char*

necto committed:
[NFC] Remove redundant [[fallthrough]] annotations

necto committed:
[NFC] remove redundant braces

necto committed:
Handle BlockExit, and humanize the BlockEntrance/BlockEdge comments

necto commented on clang/docs/analyzer/developer-docs/PerformanceInvestigation.rst:
```diff
@@ -0,0 +1,38 @@
+
+.. image:: ../images/speedscope.png
+
+On the screenshot above, under the first time ruler is the birds-eye view of the entire trace that spans a little over 8 seconds.
```
Indeed, it is not mentioned anywhere except the file name. added
bf56730224f4 mention the app name for the screenshot

necto commented on clang/docs/analyzer/developer-docs/PerformanceInvestigation.rst:
```diff
@@ -0,0 +1,38 @@
+Note: Both Chrome-tracing and speedscope tools might struggle with time traces above 100 MB in size.
+Luckily, in most cases the default max-steps boundary of 225 000 produces the traces of approximately that size
+for a single entry point.
+You should use ``-analyze-function=get_global_options`` together with ``-ftime-trace`` to narrow down analysis to a specific entry point.
```
Replaced
c55f873e8449 Less prescriptive language

necto commented on clang/lib/Analysis/ProgramPoint.cpp:
```diff
@@ -49,6 +49,127 @@ LLVM_DUMP_METHOD void ProgramPoint::dump() const {
   return printJson(llvm::errs());
 }
 
+const char *ProgramPoint::kindToStr(Kind K) {
```
Indeed, I changed the return type in
a22ed13eae22 [NFC] Return StringRef instead of a raw char*

and renamed in
299b54daa632 [NFC] rename kindToStr to getProgramPointKindName

necto commented on clang/lib/Analysis/ProgramPoint.cpp:
```diff
@@ -49,6 +49,127 @@ LLVM_DUMP_METHOD void ProgramPoint::dump() const {
+    // castAs<BlockExit>().getBlock()->getTerminatorStmt()->getBeginLoc();
+    return std::nullopt;
+  case PreStmtKind:
+    [[fallthrough]];
```
Good point, we don't. Removed
6101457c5b50 [NFC] Remove redundant [[fallthrough]] annotations

necto commented on clang/lib/Analysis/ProgramPoint.cpp:
```diff
@@ -49,6 +49,127 @@ LLVM_DUMP_METHOD void ProgramPoint::dump() const {
+std::optional<SourceLocation> ProgramPoint::getSourceLocation() const {
+  switch (getKind()) {
+  case BlockEdgeKind:
+    // return castAs<BlockEdge>().getSrc()->getTerminatorStmt()->getBeginLoc();
```
well spotted, that's a leftover from my developments.
I added a defensive implementation for BlockExit. I don't know what should be the location associated with BlockEdge and BlockEntrance, so I left them as nullopts for now.
ccbaa50c304c Handle BlockExit, and humanize the BlockEntrance/BlockEdge comments

necto commented on clang/lib/StaticAnalyzer/Core/BugReporter.cpp:
```diff
@@ -287,6 +288,35 @@ class PathDiagnosticBuilder : public BugReporterContext {
 };
 
+std::string timeTraceName(const BugReportEquivClass &EQ) {
+  if (!llvm::timeTraceProfilerEnabled()) {
```
removed
babecf833ce0 [NFC] remove redundant braces

Xazax-hun commented on clang/lib/StaticAnalyzer/Core/BugReporter.cpp:
```diff
@@ -287,6 +288,34 @@ class PathDiagnosticBuilder : public BugReporterContext {
+std::string timeTraceName(const BugReportEquivClass &EQ) {
+  if (!llvm::timeTraceProfilerEnabled())
+    return "";
+  const auto &bugReports = EQ.getReports();
```
Nit: variables should start with upper case.

necto replied:
fixed
b5792c2b135e [NFC] Uppercase variable names

Xazax-hun commented on clang/lib/StaticAnalyzer/Core/BugReporter.cpp:
```diff
@@ -287,6 +288,34 @@ class PathDiagnosticBuilder : public BugReporterContext {
+  // Must be called only when constructing non-bogus TimeTraceScope
+  assert(llvm::timeTraceProfilerEnabled());
+
+  const auto &bugReports = EQ.getReports();
```
Nit: start with upper case.

necto replied:
fixed
b5792c2b135e [NFC] Uppercase variable names

Xazax-hun commented on clang/lib/StaticAnalyzer/Core/BugSuppression.cpp:
```diff
@@ -119,6 +121,28 @@ class CacheInitializer : public DynamicRecursiveASTVisitor {
+llvm::TimeTraceMetadata getDeclTimeTraceMetadata(const Decl *DeclWithIssue) {
+  assert(DeclWithIssue);
+  assert(llvm::timeTraceProfilerEnabled());
+  std::string name = "<noname>";
```
Nit: start variables with uppercase in this function and some others below.

necto replied:
fixed
b5792c2b135e [NFC] Uppercase variable names

Xazax-hun commented on clang/lib/StaticAnalyzer/Core/CheckerManager.cpp:
```diff
@@ -134,6 +135,14 @@ static void expandGraphWithCheckers(CHECK_CTX checkCtx,
+std::string checkerScopeName(StringRef name, const CheckerBase *checker) {
+  if (!llvm::timeTraceProfilerEnabled())
+    return "";
+  std::string checkerName =
```
Nit: checkerName could be a StringRef removing one allocation.

necto replied:
fixed
b5792c2b135e [NFC] Uppercase variable names

Xazax-hun commented on clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:
```diff
@@ -358,9 +359,41 @@ class AnalysisConsumer : public AnalysisASTConsumer,
 
+std::string timeTraceScopeDeclName(StringRef FunName, const Decl *D) {
+  if (llvm::timeTraceProfilerEnabled()) {
+    if (const NamedDecl *ND = dyn_cast<NamedDecl>(D)) {
```
Braces redundant here.

necto replied:
ac8707b0db5e [NFC] Remove redundant braces

Xazax-hun commented on clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:
```diff
@@ -358,9 +359,41 @@ class AnalysisConsumer : public AnalysisASTConsumer,
+    const auto &SM = D->getASTContext().getSourceManager();
+    std::string DeclName = AnalysisDeclContext::getFunctionName(D);
+    return llvm::TimeTraceMetadata{
+        DeclName, SM.getFilename(Loc).str(),
```
Could move `DeclName`.

necto replied:
Moved, thanks
a63907b742f6 [NFC] Move some strings

Xazax-hun commented on clang/lib/StaticAnalyzer/Core/BugSuppression.cpp:
```diff
@@ -119,6 +121,28 @@ class CacheInitializer : public DynamicRecursiveASTVisitor {
+  const auto &SM = DeclWithIssue->getASTContext().getSourceManager();
+  auto line = SM.getPresumedLineNumber(DeclWithIssue->getBeginLoc());
+  auto fname = SM.getFilename(DeclWithIssue->getBeginLoc());
+  return llvm::TimeTraceMetadata{name, fname.str(), static_cast<int>(line)};
```
Could move name here.

Xazax-hun commented on clang/lib/StaticAnalyzer/Core/BugReporter.cpp:
```diff
@@ -287,6 +288,34 @@ class PathDiagnosticBuilder : public BugReporterContext {
+  std::string File = "";
+  if (const auto *Entry = Loc.getFileEntry())
+    File = Entry->tryGetRealPathName().str();
+  return {BT.getCheckerName().str(), File,
```
Could move File here.

necto replied:
Moved, thanks
a63907b742f6 [NFC] Move some strings

Xazax-hun reviewed (approved):
Some nits inline for style and removing allocations. Overall looks good for me.

necto committed:
[NFC] Uppercase variable names

necto committed:
[NFC] Remove redundant braces

necto committed:
[NFC] Move some strings

Xazax-hun commented on clang/lib/StaticAnalyzer/Core/CheckerManager.cpp:
```diff
@@ -365,13 +388,22 @@ namespace {
 
+  llvm::TimeTraceMetadata getTimeTraceBindMetadata(SVal val) {
+    assert(llvm::timeTraceProfilerEnabled());
+    std::string name;
```
Some leftover lowercase variable names here.

necto replied:
Sorry for overlooking these
4cc3be9658c4 [NFC] Uppercase some variable names

Xazax-hun commented on clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:
```diff
@@ -179,8 +181,40 @@ bool CoreEngine::ExecuteWorkList(const LocationContext *L, unsigned MaxSteps,
+                                                 const ProgramPoint &Loc) {
+  // If time-trace profiler is not enabled, this function is never called.
+  assert(llvm::timeTraceProfilerEnabled());
+  std::string str;
```
Some lowercase names here as well.

necto replied:
Sorry for overlooking these
4cc3be9658c4 [NFC] Uppercase some variable names

Xazax-hun reviewed (approved):
There are some lowercase variable names still, but otherwise it still looks good to me. Feel free to commit after those are fixed.

necto committed:
[NFC] Undo whitespace change

necto committed:
[NFC] Uppercase some variable names

balazs-benics-sonarsource reviewed (approved):
I had a look at the PR, and it looks awesome.
Could you please update the attached speedscope image?
It looks like it's out of sync with the implementation, for example if you look at the "Loc PostStmt { ... stuff here ...}" , it appears to include the ProgramPoint dump, and according to your implementation it should only have "Loc <ProgramPointKindName>". Am I right?

necto commented:
> I had a look at the PR, and it looks awesome. Could you please update the attached speedscope image? It looks like it's out of sync with the implementation, for example if you look at the "Loc PostStmt { ... stuff here ...}" , it appears to include the ProgramPoint dump, and according to your implementation it should only have "Loc ". Am I right?

speedscope puts the "details" onto the scope labels in addition to their name, unlike chrome-tracing that only displays the name, so that's the up-to-date display

necto committed:
Remove full ProgramPoint dumps from the traces to avoid unbounded labels

necto committed:
put ftime-trace CL arguments on a dedicatede line

necto commented:
> > I had a look at the PR, and it looks awesome. Could you please update the attached speedscope image? It looks like it's out of sync with the implementation, for example if you look at the "Loc PostStmt { ... stuff here ...}" , it appears to include the ProgramPoint dump, and according to your implementation it should only have "Loc ". Am I right?
> 
> speedscope puts the "details" onto the scope labels in addition to their name, unlike chrome-tracing that only displays the name, so that's the up-to-date display

@steakhal, As agreed out-of-band, I removed the potentially ounbounded ProgramPoint dumps from the trace, and replace them with just the statement kind for `StmtKind` program points. I've updated the doc accordingly, please take another look.

balazs-benics-sonarsource commented on clang/docs/analyzer/developer-docs/PerformanceInvestigation.rst:
```diff
@@ -14,17 +14,26 @@ such as `Perf <https://perfwiki.github.io/main/>`_ and `Callgrind <https://valgr
+.. code-block:: bash
+   :caption: Clang Static Analyzer invocation to generate a time trace of string.c analysis.
+
+   clang -cc1 -internal-isystem ./build/asserts/lib/clang/21/include \
```
You can drop the `./build/asserts/lib/clang/21/include` include path because the test doesn't use any includes, and currently this hardcodes the clang-21 version otherwise.

necto replied:
Nice!
d5851246fe2a Remove unnecessary include path

balazs-benics-sonarsource reviewed (commented):
Looks even better now. Thanks.
I found one final nit remaining.

necto committed:
Remove unnecessary include path

balazs-benics-sonarsource commented:
LGTM, I'll merge this PR once the premerge checks are green. Should be ready in a couple of hours. Thanks for the PR again!

balazs-benics-sonarsource commented on clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:
```diff
@@ -179,8 +181,41 @@ bool CoreEngine::ExecuteWorkList(const LocationContext *L, unsigned MaxSteps,
+
+void CoreEngine::dispatchWorkItem(ExplodedNode *Pred, ProgramPoint Loc,
+                                  const WorkListUnit &WU) {
+  llvm::TimeTraceScope tcs{timeTraceScopeName(Loc), [Loc, Pred]() {
+                             return timeTraceMetadata(Pred, Loc);
+                           }};
```
It took me a while to correlate the name `Loc PostStmt` with this place. I wonder if we should use `work item` as a disambiguation.
